intercept/src/intercept.cpp

```diff
@@ -7244,13 +7304,212 @@ void CLIntercept::setKernelArgUSMPointer(
     }
 }
 
+void CLIntercept::dumpKernelSource(cl_kernel kernel, 
+                                   uint64_t enqueueCounter,
+                                   bool byKernelName )
+{
+    std::lock_guard<std::mutex> lock(m_Mutex);
+    std::string fileNamePrefix = "";
+    OS().GetDumpDirectoryName( sc_DumpDirectoryName, fileNamePrefix );
+    fileNamePrefix += "/Replay/Enqueue_";
+    if (byKernelName)
+        fileNamePrefix += getShortKernelName(kernel);
+    else
+        fileNamePrefix += std::to_string(enqueueCounter);
+    fileNamePrefix += "/";
+    OS().MakeDumpDirectories( fileNamePrefix );
+
+    // Get the cl_program from the cl_kernel, then extract kernel source code from the cl_program
+    CLIntercept* pIntercept = GetIntercept();
+
+    cl_program tmp_program;
+    pIntercept->dispatch().clGetKernelInfo(kernel, CL_KERNEL_PROGRAM, sizeof(cl_program), &tmp_program, nullptr);
+
+    try
+    {
+        std::string source = m_SourceStringMap.at(tmp_program);
+        std::string fileName = fileNamePrefix + "kernel.cl";
+        std::ofstream output;
+        output.open(fileName.c_str(), std::ios::out | std::ios::binary );
+        output.write(source.c_str(), source.length());
+        return;
+    } 
+    catch (std::exception const& excep)
+    {
+        (void)excep;
+    }
+
+    log("[[Warning]]: Kernel source is not available! Make sure that the kernel is compiled from source (and is not cached)\n");
+    log("Now will try to output binaries, these probably won't work on other platforms!\n");
+    
+    cl_uint num_devices;
+    clGetProgramInfo(tmp_program, CL_PROGRAM_NUM_DEVICES, sizeof(cl_uint), &num_devices, nullptr);
+
+    // Grab the device ids
+    cl_device_id* devices = new cl_device_id[num_devices];
+    clGetProgramInfo(tmp_program, CL_PROGRAM_DEVICES, num_devices * sizeof(cl_device_id), devices, 0);
+
+    std::string knlName = getShortKernelName(kernel);
+    for (unsigned device = 0; device < num_devices; ++device)
+    {
+        auto binaries = m_DeviceBinaryMap[devices[device]];
+        for (unsigned binary = 0; binary != binaries.size(); ++binary)
+        {
+            auto pos = std::search(binaries[binary].data(), binaries[binary].data() + binaries[binary].size(), knlName.data(), knlName.data() + knlName.length());
+            if (pos != binaries[binary].data() + binaries[binary].size())
```

Comment:
This is probably safe, but strictly speaking the program binary does not need to contain the kernel name :smile: